conda/gateways/repodata/__init__.py

```diff
@@ -656,6 +710,251 @@ def timeout(self):
 
 
+class RepodataFetch:
+    """
+    Combine RepodataCache and RepoInterface to provide subdir_data.SubdirData()
+    with what it needs.
+
+    Provide a variety of formats since some ``RepoInterface`` have to
+    ``json.loads(...)`` anyway, and some clients don't need the Python data
+    structure at all.
+    """
+
+    cache_path_base: Path
+    channel: Channel
+    repodata_fn: str
+    url_w_subdir: str
+    url_w_credentials: str
+    repo_interface_cls: Any
+
+    def __init__(
+        self,
+        cache_path_base: Path,
+        channel: Channel,
+        repodata_fn: str,
+        *,
+        repo_interface_cls,
+    ):
+        self.cache_path_base = cache_path_base
+        self.channel = channel
+        self.repodata_fn = repodata_fn
+
+        self.url_w_subdir = self.channel.url(with_credentials=False) or ""
+        self.url_w_credentials = self.channel.url(with_credentials=True) or ""
+
+        self.repo_interface_cls = repo_interface_cls
+
+    def fetch_latest_parsed(self) -> tuple[dict, RepodataState]:
+        """
+        Retrieve parsed latest or latest-cached repodata as a dict; update
+        cache.
+
+        :return: (repodata contents, state including cache headers)
+        """
+        parsed, state = self.fetch_latest()
+        if isinstance(parsed, str):
+            return json.loads(parsed), state
+        else:
+            return parsed, state
+
+    def fetch_latest_path(self) -> tuple[Path, RepodataState]:
+        """
+        Retrieve latest or latest-cached repodata; update cache.
+
+        :return: (pathlib.Path to uncompressed repodata contents, RepodataState)
+        """
+        _, state = self.fetch_latest()
+        return self.cache_path_json, state
+
+    @property
+    def url_w_repodata_fn(self):
+        return self.url_w_subdir + "/" + self.repodata_fn
+
+    @property
+    def cache_path_json(self):
+        return Path(
+            str(self.cache_path_base)
+            + ("1" if context.use_only_tar_bz2 else "")
+            + ".json"
+        )
+
+    @property
+    def cache_path_state(self):
+        """
+        Out-of-band etag and other state needed by the RepoInterface.
+        """
+        return Path(
+            str(self.cache_path_base)
+            + ("1" if context.use_only_tar_bz2 else "")
+            + CACHE_STATE_SUFFIX
+        )
+
+    @property
+    def repo_cache(self) -> RepodataCache:
+        return RepodataCache(self.cache_path_base, self.repodata_fn)
+
+    @property
+    def _repo(self) -> RepoInterface:
+        """
+        Changes as we mutate self.repodata_fn.
+        """
+        return self.repo_interface_cls(
+            self.url_w_credentials,
+            repodata_fn=self.repodata_fn,
+            cache_path_json=self.cache_path_json,
+            cache_path_state=self.cache_path_state,
+            cache=self.repo_cache,
+        )
+
+    def fetch_latest(self) -> tuple[dict | str, RepodataState]:
```

Comment:
This code is very similar to what was removed from `subdir_data.py`

Reply:
gotcha, so in this PR we will refrain from additional changes here 👍🏼